### PR TITLE
Refine deadlock

### PR DESCRIPTION
**Pull request by youjiali1995**
## What have you changed? (mandatory)
* The New leader won't get `wait_for_entries` from followers.
* Added duration of `monitor_membership_change` metric.
* Check leadership in detect_rpc.

## What are the type of the changes? (mandatory)
- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
Manual tests. The new leader with an empty detect table doesn't affect functionality.

## Does this PR affect documentation (docs) or release note? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
Yes. https://github.com/pingcap/tidb-ansible/pull/781

### CODE REVIEW

**Comment by youjiali1995** [comment]
@coocood @zhangjinpeng1987 @AndreMouche PTAL

File: src/storage/lock_manager/client.rs [hunk 1]
```diff
@@ -43,6 +43,8 @@ impl Client {
         }
     }
 
+    // TODO: remove it
```
Comment (coocood) [inline src/storage/lock_manager/client.rs]: Why not remove it now?

File: src/storage/lock_manager/deadlock.rs [hunk 2]
```diff
@@ -222,126 +230,42 @@ impl Inner {
...
                 }
-            });
-            self.failed.borrow_mut().clear();
+                self.leader_info.replace((leader_id, leader_addr));
+            }
         }
     }
 
-    fn become_leader(&mut self, handle: &Handle) {
+    fn become_leader(&mut self) {
```
Comment (coocood) [inline src/storage/lock_manager/deadlock.rs]: The only difference between `become_leader` and `become_follower` is log message.
I think we can inline the two functions into `change_role_if_needed`.

**Comment by coocood** [comment]
LGTM

File: src/storage/lock_manager/deadlock.rs [hunk 3]
```diff
@@ -198,20 +195,25 @@ impl Inner {
...
-            addrs: Rc::new(RefCell::new(HashMap::new())),
-            leader_client: None,
-            recving: Rc::new(RefCell::new(HashMap::new())),
-            failed: Rc::new(RefCell::new(HashSet::new())),
             max_ts: 0,
         }
     }
 
+    fn leader_id(&self) -> u64 {
+        self.leader_info.as_ref().unwrap().0
```
Comment (zhangjinpeng1987) [inline src/storage/lock_manager/deadlock.rs]: How to guarantee that leader_info is Some(xx)?

File: src/storage/lock_manager/deadlock.rs [hunk 4]
```diff
@@ -402,21 +309,48 @@ impl Detector {
...
+                Some(ref addr) if addr.is_ok() => {
+                    inner
+                        .borrow_mut()
+                        .change_role_if_needed(leader_id, addr.as_ref().unwrap().to_owned());
+                    Ok(())
+                }
+                _ => Err(box_err!("failed to resolve leader address")),
+            };
+        }
+        warn!("leader not found");
```
Comment (zhangjinpeng1987) [inline src/storage/lock_manager/deadlock.rs]: Miss else?

Comment (youjiali1995) [reply src/storage/lock_manager/deadlock.rs]: L323 has return.

File: src/storage/lock_manager/deadlock.rs (hunk 3 above)
Comment (youjiali1995) [reply src/storage/lock_manager/deadlock.rs]: Only `is_leader()` and `reconnect_leader()` call these functions. `is_leader()` has checked `leader_info.is_some()` and L410 has checked it too.

**Pull request merged**